### ISSUE TITLE
change tab disappears when generic object plugin is activated

### ISSUE THREAD

**Issue by satyan01** [pluginsGLPI/genericobject#114]
Hello

change tab disappears in ticket when generic object plugin is activated

Please Help

Thank you

glpi9.1 - plugin 2.4.0

**Comment by satyan01** [comment]
Hello

Do you have a solution please  ? 6 days ago • edited

**Comment by satyan01** [comment]
Hello

I can't use the plugin please help

**Comment by btry** [comment]
Hi

Please check in glpi/files/_log : isthere relevant entries in php-errors.log or sql-errors.log ?

You also can try to reproduce the issue with a super admin account, debug mode enabled and see if text with red background appears.

**Comment by btry** [comment]
I'm testing with GLPi 9.1 (branch 9.1/bugfixes), Generic Object 2.4.0 (release)

I tried to create an new object, but I don't reproduce yet. 

![image](https://cloud.githubusercontent.com/assets/14139801/20176675/2b017fb2-a749-11e6-9e7c-d53112f5b44b.png)

![image](https://cloud.githubusercontent.com/assets/14139801/20176689/3e89fbae-a749-11e6-84da-6a35fb62c6a0.png)

If you got a testing environment, disable one object, and check if the tickets still miss their "change" tab. Repeat the process until the tab appears again or you disabled all your objects.

**Comment by satyan01** [comment]
Hello

No sql log or php log relevant.

![image](https://cloud.githubusercontent.com/assets/15144026/20177625/fe120eea-a74d-11e6-87cf-c0b4c2f6c26c.png)

![image](https://cloud.githubusercontent.com/assets/15144026/20177637/158dceb0-a74e-11e6-89d9-9e887730faa4.png)

**Comment by satyan01** [comment]
when I disable my generic object "key" it comes back

**Comment by satyan01** [comment]
![image](https://cloud.githubusercontent.com/assets/15144026/20177699/54ccde22-a74e-11e6-938b-8fe1162bf1fd.png)

**Comment by btry** [comment]
Great !

I'll investigate soon with a similar object. Thank you.

**Comment by satyan01** [comment]
It works until I create a second generic object

**Comment by satyan01** [comment]
So it does not come from the generic object, it comes from second creation

**Comment by btry** [comment]
I tried, and I don't reproduce. Note my objects are empty. 

I planned to search how the behaviours may help to reproduce the issue.

Can you show a screenshot of your second object ?

**Comment by satyan01** [comment]
![image](https://cloud.githubusercontent.com/assets/15144026/20178977/fdf32a2e-a753-11e6-860b-52beccf96ccb.png)

**Comment by satyan01** [comment]
I have tried with empty object and at each time I create a new one and activate it change disappears

**Comment by satyan01** [comment]
If I create a third object it works correctly

**Comment by btry** [comment]
I suspect I did not activate t he first object in my last try.

I reproduce now, with the following object 

![image](https://cloud.githubusercontent.com/assets/14139801/20179312/d2e8358e-a755-11e6-8d90-30ca8d59a657.png)

I created a second object, tried to name it as yours, and I finally found it is not necessary.

If I disable **Key** the tabs appears again

**Comment by satyan01** [comment]
so it is only a second one that makes the error

**Comment by btry** [comment]
I'll investigate first with the current state of my instance of GLPi. 

I'll invite you to  test the fix after I found it.

**Comment by satyan01** [comment]
When I have three objects and I unable the second one it works

**Comment by satyan01** [comment]
again change disappear

**Comment by btry** [comment]
without any action on generic objects ?

**Comment by satyan01** [comment]
no

**Comment by satyan01** [comment]
I have just let objects active, when technicians have refresh their page they lost change tab

**Comment by satyan01** [comment]
after unable it tab change reappears

**Comment by btry** [comment]
Hi

I think I found the problem.

However, it seems the name of the object has something to do with the bug, maybe also the order of loading the classes when rendering a page. 

Here is the patch : 
https://github.com/btry/genericobject/tree/114_tabs_disappear_when_objects_enabled

and a direct link to the ZIP file : https://github.com/btry/genericobject/archive/114_tabs_disappear_when_objects_enabled.zip

Please test it exclusively on a testing environment, and use the same objects as those on your production environment because names seems to affect how the bug occurs.

**Comment by btry** [comment]
Oh I forgot to mention : you *must* recreate the classes of your objects !

there is a CLI php script in the folder tools of the plugin named **check_classes.php** to do this easily.

An other way is to use the button in the **Debug** tab of each object.

**Comment by satyan01** [comment]
It seems to work, I'll keep you informed next days.

What was it ?

**Comment by satyan01** [comment]
Thank you very much for your work !!!

**Comment by btry** [comment]
Hi

To me this was a problem of inheritance from a class used as base class for all types of objects in GLPi. This class is also a base of the types you create with the plugin.

the plugin defines the name of the rights of your types, but the property is statically (see OOP in PHP) defined in GLPi. Then the value affects GLPi itself and broke the rights checks before displaying some tabs.

However, the bug should occur as soon a generic object is enabled, but I found it occurred only if a generic object has a short name ending with the character *y*. (your object **Key**). I did not find why these details affect the occurence of the bug : the related source code is very short and trivial ...

If you're interested see the file object.class.php in the commit https://github.com/btry/genericobject/commit/fdabc4b7478f37b0f78e93a17f5522c98c4af101#diff-c7d2c155f00fc5787c51bca2ac6a2d18

I changed two lines, but only the first one triggered the bug. The second did not had the same negative impact, hereas it should (from my point of view).

The patch redefines and uses the static property at the deepest inheritance level of your objects to stop unwanted propagation of the affectation to other (parent) classes.
